docs/decisions/infra/0005-database-migration-architecture.md

```diff
@@ -0,0 +1,129 @@
+
+#### Infrastructure Required
+
+- Lambda function to run the migration code
+- Log group for storing logs
+- IAM role for the Lambda to use
+- Some other infrastructure to store the state of the migration (S3, EC2, ECS)
+- Any associated IAM roles for the state-storing infrastructure to use
+- Security group that allows connection to the database
+- IAM role that Github can assume to run the task
+- (optional) A queueing method for the Lambda to feed off of, like SQS
+
+#### Pros
+
+Lambda functions can be very cost effective and fast, as well as simple to use.
+
+Lambdas are low maintenance, and don't stick around after the task.
+
+Even though this solution requires additional support, the compute portion of the task run by AWS Lambda is very simple.
+
+#### Cons
+
+By nature, database migrations need to keep track of state--changes between the previous version, the next version, and at what point the migration is currently at. Lambda functions are stateless by design, so an additional component will be needed to keep track of state.
+
+Lambdas have a comparatively short running time. Either find a way to split up the migrations into small tasks, or extend the running time of the Lambda. This might not scale well with large changesets.
+
+
+### Execute via an ECS Fargate Task
+
+In this method, the migration is ran by an ECS task that spins up in a cluster.
+This task will be destroyed once the migration is completed, so it will not need to be
+a long running service.
+
+This method gets the latest configuration by creating a Dockerized image from the
+workflow and pushing it to AWS ECR. This image has a function that will be called
+from the ECS task's configuration JSON.
+
+#### Infrastructure Required
+
+ - ECS cluster to run the task in, which can be the same cluster as the application
+ - Log group for storing logs
+ - IAM role for the task to use
+ - (Optional) IAM policy with permission to connection if using IAM authentication on the database
+ - (Optional) ECR to store the image. This is dependent on the application used
+ - Security group that allows connection to the database
+ - IAM role that GitHub can assume to run the task
+
+#### Pros
+
+Depending on the application, you can overload the ECS task configuration to run
+the migration command and use the same app image, so there is no need to maintain
+a different image. This allows you to run the same code locally as in AWS to try
+and catches errors before merging. (An example of this is the flask application)
+
+Cheaper to run compared to Lambda, ~$0.015/hour vs ~$0.06/hour
+
+#### Cons
+
+The only way to re-run migrations if there is a failure, is to re-run the workflow from GitHub.
+
+Start-up time is slower compared to Lambda
```

Comment:
how much slower? can we put some rough numbers / orders of magnitude for startup times of each approach?

Reply:
It depends on the provisioning step in environment. Lambdas have the option to setup the provisioning outside of being ran, so that when it is triggered, the function is able to just run. ECS tasks will have to do that provisioning each time it runs, so it takes somewhat longer to run. However, the costs might balance this time difference out since Lambdas is ~2.5x more expensive. Updated the ADR with more of a blurb on this and how to set it up

Reply:
i think it can be useful to put the cost information in their but i would weigh the cost info as negligible compared to the cost/benefit of maintenance efficiency for engineers. database (RDS) costs dwarf all other compute costs, and the cost of engineers dwarf all AWS costs, so 2.5x difference is pretty negligible for something that only runs a few seconds per deploy